### PR TITLE
feat(memory): Separate total memory configurations for FE and Compactor

### PR DESCRIPTION
**Pull request by Li0k**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

part of https://github.com/risingwavelabs/risingwave/issues/19245

This PR introduces a separate total memory configuration for FE and Compactor.

In standalone deployment mode, the total_memory of different nodes should be set to achieve memory isolation.  (Not included in this PR)

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

**Review by hzxa21 — APPROVED** [review]

**Review by chenzl25 — APPROVED** [review]
LGTM

File: src/frontend/src/lib.rs [hunk 1]
```diff
@@ -159,6 +160,10 @@ pub struct FrontendOpts {
         default_value = "./secrets"
     )]
     pub temp_secret_file_dir: String,
+
+    /// Total available memory for the frontend node in bytes. Used by both computing and storage.
+    #[clap(long, env = "RW_FE_TOTAL_MEMORY_BYTES", default_value_t = default_fe_total_memory_bytes())]
```
Comment (chenzl25) [inline src/frontend/src/lib.rs:165]: change `RW_FE_TOTAL_MEMORY_BYTES` to `RW_FRONTEND_TOTAL_MEMORY_BYTES`?

File: src/frontend/src/lib.rs [hunk 2]
```diff
@@ -159,6 +160,10 @@ pub struct FrontendOpts {
         default_value = "./secrets"
     )]
     pub temp_secret_file_dir: String,
+
+    /// Total available memory for the frontend node in bytes. Used by both computing and storage.
```
Comment (kwannoel) [inline src/frontend/src/lib.rs:164]:
```suggestion
    /// Total available memory for the frontend node in bytes. Used by frontend.
```

File: src/storage/src/hummock/store/hummock_storage.rs [hunk 3]
```diff
@@ -724,7 +724,6 @@ impl HummockStorage {
     }
 
     /// Used in the compaction test tool
-    #[cfg(any(test, feature = "test"))]
     pub async fn update_version_and_wait(&self, version: HummockVersion) {
         use tokio::task::yield_now;
         let version_id = version.id;
```
Comment (kwannoel) [inline src/storage/src/hummock/store/hummock_storage.rs:729]: Are changes in this file expected?

File: src/storage/compactor/src/server.rs [hunk 4]
```diff
@@ -81,7 +82,7 @@ pub async fn prepare_start_parameters(
         &system_params_reader,
         &storage_memory_config,
     )));
-    let non_reserved_memory_bytes = (system_memory_available_bytes() as f64
+    let non_reserved_memory_bytes = (compactor_opts.compactor_total_memory_bytes as f64
```
Comment (kwannoel) [inline src/storage/compactor/src/server.rs:85]: Compactor opts provide an instance of `compactor_total_memory_bytes`, but inside `RwConfig`, there's a separate value:
```
    pub compactor_memory_limit_mb: Option<usize>,
```

What's the difference between these configurations? Will the one in RwConfig always be ignored, if so can we remove it?

**Review by kwannoel — APPROVED** [review]
Generally LGTM, thanks for your work :)

File: src/storage/src/hummock/store/hummock_storage.rs (hunk 3 above)
Comment (Li0k) [reply src/storage/src/hummock/store/hummock_storage.rs:729]: right

File: src/storage/compactor/src/server.rs (hunk 4 above)
Comment (Li0k) [reply src/storage/compactor/src/server.rs:85]: Good question:

I found that compactor_memory_limit_mb is only used in embedding compactor mode and occupies part of the CN's memory (you can refer to the storage_memory_config function).

I think this is different from standalone deployment mode, I didn't change it in this PR, maybe you have a better idea?

Comment (kwannoel) [reply src/storage/compactor/src/server.rs:85]: > Good question:
> I found that compactor_memory_limit_mb is only used in embedding compactor mode and occupies part of the CN's memory (you can refer to the storage_memory_config function).
> I think this is different from standalone deployment mode, I didn't change it in this PR, maybe you have a better idea?

Got your point. It's out of scope for this PR then. We can just rename it.
IIRC embedded compactor is not used for standalone deployment.

**Comment by hzxa21** [comment]
Should we cherry pick this to 2.1?

File: src/frontend/src/lib.rs (hunk 2 above)
Comment (fuyufjh) [inline src/frontend/src/lib.rs:164]: Does storage run in frontend? IIRC, even in local execution mode, the `TableScan` is pushed to compute nodes.

Comment (hzxa21) [reply src/frontend/src/lib.rs:164]: I think this comment is not resolved.

**Pull request merged**